compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -300,60 +302,84 @@ fn negative_impl<'cx, 'tcx>(
     debug!("negative_impl(impl1_def_id={:?}, impl2_def_id={:?})", impl1_def_id, impl2_def_id);
     let tcx = selcx.infcx().tcx;
 
-    // create a parameter environment corresponding to a (placeholder) instantiation of impl1
-    let impl1_env = tcx.param_env(impl1_def_id);
-    let impl1_trait_ref = tcx.impl_trait_ref(impl1_def_id).unwrap();
-
     // Create an infcx, taking the predicates of impl1 as assumptions:
     tcx.infer_ctxt().enter(|infcx| {
-        // Normalize the trait reference. The WF rules ought to ensure
-        // that this always succeeds.
-        let impl1_trait_ref = match traits::fully_normalize(
-            &infcx,
-            FulfillmentContext::new(),
-            ObligationCause::dummy(),
-            impl1_env,
-            impl1_trait_ref,
-        ) {
-            Ok(impl1_trait_ref) => impl1_trait_ref,
-            Err(err) => {
-                bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
-            }
-        };
+        // create a parameter environment corresponding to a (placeholder) instantiation of impl1
+        let impl1_env = tcx.param_env(impl1_def_id);
+
+        if let Some(impl1_trait_ref) = tcx.impl_trait_ref(impl1_def_id) {
```

Comment:
I was suggesting on Zulip that we have

```rust
pub enum ImplHeader<'tcx> {
    Trait(TraitRef<'tcx>),
    Inherent(Ty<'tcx>)
}
```

and then a method on tcx that returns one or the other. You could then implement foldable and the like for that. This would allow you to just unify them.

Reply:
Yeah, that's a cool way to do what I was saying in https://github.com/rust-lang/rust/pull/95082#discussion_r830206485

Reply:
I've added an enum in https://github.com/rust-lang/rust/pull/95082/commits/0cd03c917cd728844048edb41a3d30d414214ad8 but I'm pretty sure that commit is not what you want. The problem is that we need to process those 2 def_ids and return `T: ToTrace` and an iter for obligations. I could have a function that returns those and then inline what right now lives in `equate`.